Northstar.CustomServers/mod/scripts/vscripts/mp/_stats.nut

```diff
@@ -12,67 +14,1083 @@ global function PreScoreEventUpdateStats
+}
+
+void function UpdateTitanDamageStat( entity attacker, float savedDamage, var damageInfo )
+{
+	if ( !IsValid( attacker ) )
+		return
+		
+	Stats_IncrementStat( attacker, "titan_stats", "titanDamage", GetTitanCharacterName( attacker ), savedDamage )
+}
+
+void function UpdateTitanWeaponDamageStat( entity attacker, float savedDamage, var damageInfo )
+{
+	if ( !IsValid( attacker ) )
+		return
+
+	string weaponName = GetPersistenceRefFromDamageInfo( damageInfo )
+	if ( weaponName == "" )
+		return
+
+	Stats_IncrementStat( attacker, "weapon_stats", "titanDamage", weaponName, savedDamage )
+}
+
+void function UpdateTitanCoreEarnedStat( entity player, entity titan, int count = 1 )
+{
+	if ( !IsValid( player ) )
+		return
+
+	if ( !IsValid( titan ) )
+		return
+
+	Stats_IncrementStat( player, "titan_stats", "coresEarned", GetTitanCharacterName( titan ), count.tofloat() )
+}
+
+void function PreScoreEventUpdateStats( entity attacker, entity ent )
+{
+	// used to track kill streaks ending i think ( that stuff gets reset during score event update )
+}
+
+void function PostScoreEventUpdateStats( entity attacker, entity ent )
+{
+	if ( !attacker.IsPlayer() )
+		return
+	// used to track kill streaks starting maybe
+	if ( attacker.s.currentKillstreak == KILLINGSPREE_KILL_REQUIREMENT )
+	{
+		// killingSpressAs_<chassis>
+		if ( attacker.IsTitan() )
+			Stats_IncrementStat( attacker, "kills_stats", "killingSpressAs_" + GetTitanCharacterName( attacker ), "", 1.0 )
+
+		entity weapon = attacker.GetActiveWeapon()
+		// I guess if you dont have a valid active weapon when you get awarded a killing spree
+		// you dont get the stat. Too bad!
+		if ( !IsValid( weapon ) )
+			return
+		Stats_IncrementStat( attacker, "weapon_kill_stats", "killingSprees", weapon.GetWeaponClassName(), 1.0 )
+	}
+}
+
+void function Stats_OnPlayerDidDamage( entity victim, var damageInfo )
+{
+	// try and get the player
+	entity attacker = DamageInfo_GetAttacker( damageInfo )
+	// get the player from their titan
+	if ( attacker.IsTitan() && IsPetTitan( attacker ) )
+		attacker = attacker.GetTitanSoul().GetBossPlayer()
+
+	if ( !attacker.IsPlayer() )
+		return
+
+	string weaponName = GetPersistenceRefFromDamageInfo( damageInfo )
+	if ( weaponName == "" )
+		return
+
+	Stats_IncrementStat( attacker, "weapon_stats", "shotsHit", weaponName, 1.0 )
+
+	if ( DamageInfo_GetCustomDamageType( damageInfo ) & DF_CRITICAL )
+		Stats_IncrementStat( attacker, "weapon_stats", "critHits", weaponName, 1.0 )
+	if ( DamageInfo_GetCustomDamageType( damageInfo ) & DF_HEADSHOT )
+		Stats_IncrementStat( attacker, "weapon_stats", "headshots", weaponName, 1.0 )
+}
+
+void function Stats_IncrementStat( entity player, string statCategory, string statAlias, string statSubAlias, float amount )
 {
+	if ( !IsValidStat( statCategory, statAlias, statSubAlias ) )
+	{
+		printt( "invalid stat: " + statCategory + " : " + statAlias + " : " + statSubAlias )
```

Comment:
```suggestion
		printt( format( "invalid stat: %s : %s : %s", statCategory, statAlias, statSubAlias ) )
```
string concats are heavier than in other languages because strings are immutable

Reply:
I don't think it really matters, this is basically an error message and shouldn't get hit